### DISCUSSION TITLE
Time traveler (this specific map) not working

### DISCUSSION THREAD

**Discussion by Euro20179** [ppy/osu#17062] [Q&A] [answered]
Not really sure where I can ask for help so I'm asking here.

For some people on lazer this map works, for other people it does not.
For me it does not work on Linux, it also does not work for my friend on windows 10.
However it seems to work for other people on Linux.

I've tried redownloading the map it doesn't work.

(If you're wondering why i care so much about this, it's because i think the map is cool and it works perfectly fine for me in stable on linux through wine)

I have no idea what is and is not relevent, so here's the [runtime log](https://gist.githubusercontent.com/Euro20179/5f1c4d0be3a45329349493e8bb5783d3/raw/61d036b93cba85c7b68b6ecadcb7d86dc7406f2a/runtime.log)

I also played the map during this session at 03:25:00 ish (basically at the end of the log)

EDIT:
 just occured to me i should be more specific about what is not working
basically the parts where your supposed to hit the notes backwards doesnt work, as in the notes (that are probably on the storyboard) don't appear. However the end of the song where the skin changes works fine.

**Comment by bdach** [comment]
please explain which section in particular you mean. video would also be appreciated.

**Comment by Walavouchey** [comment]
can reproduce on windows 10

[beatmap in question](https://osu.ppy.sh/beatmapsets/1236988#osu/2573493)

the problem is that the storyboarded approach circles and other skin elements don't appear

the sprite file name is declared as `"approachcircle.png"`, with no such file in the mapset folder. this is because it's supposed to use the skin sprite, which is apparently not working on lazer (`UseSkinSprites: 1` is declared in the .osu file)

note that there is such a file in the `old_skin` folder, but that one doesn't have low opacity and is only used for the nostalgia (and lazer) section of the storyboard

[logs.zip](https://github.com/ppy/osu/files/8191396/logs.zip)

liveplay at 0% background dim (the correct way to play the game), see 1:31:

https://user-images.githubusercontent.com/36758269/156899822-6a217ce5-c78a-415c-b0a9-5f29924df688.mp4

stable for comparison, see 1:24:

https://user-images.githubusercontent.com/36758269/156899206-71b722bc-42c4-4bea-9794-f05aabcc715e.mp4

**Reply by bdach** [reply]
thanks for the information. i will however note that the assertion that `UseSkinSprites` "is apparently not working on lazer" is not wholly true.

the reason why the skin sprites are failing to appear on this particular map is because they appear in the storyboard with a `.png` extension at the end of the filename, which `SpriteComponent` does not handle correctly:

https://github.com/ppy/osu/blob/4de66bb1c6e053a948b968d85905c0dda95f8a2b/osu.Game/Skinning/SkinnableSprite.cs#L36-L44

the following crude diff fixes this particular case:

```diff
diff --git a/osu.Game/Storyboards/Storyboard.cs b/osu.Game/Storyboards/Storyboard.cs
index c4864c0334..173b36676d 100644
--- a/osu.Game/Storyboards/Storyboard.cs
+++ b/osu.Game/Storyboards/Storyboard.cs
@@ -103,7 +103,12 @@ public Drawable CreateSpriteFromResourcePath(string path, TextureStore textureSt
                 drawable = new Sprite { Texture = textureStore.Get(storyboardPath) };
             // if the texture isn't available locally in the beatmap, some storyboards choose to source from the underlying skin lookup hierarchy.
             else if (UseSkinSprites)
+            {
+                if (path.EndsWith(@".png", StringComparison.Ordinal))
+                    path = path.Substring(0, path.Length - 4);
+
                 drawable = new SkinnableSprite(path);
+            }
 
             return drawable;
         }
```

the real PR should probably check more extensions. not sure how many. it's a bit sketch.

**Reply by Walavouchey** [reply]
i'm not very versed in the workings of this, but wouldn't it be better if `SkinnableSprite` could take the path with the extension? rather, wouldn't all cases be fixed just by stripping any extension? i presume skin sprites are not comparing against file extensions because an image in a skin can have a wide range of extensions.

and fyi the only thing i asserted was that "it's supposed to use the skin sprite, which is apparently not working on lazer" for this particular instance, while presenting all potentially relevant information

**Reply by bdach** [reply]
> i presume skin sprites are not comparing against file extensions because an image in a skin can have a wide range of extensions

that would be correct.

**Reply by frenzibyte** [reply]
I haven't digged deep into this, but wouldn't the sprite resource be found *with* the extension? looking at `ResourceStore`, [it considers not adding an extension to the given name as a candidate filename](https://github.com/ppy/osu-framework/blob/efcbe56398543652089fbebc96d1fda30f9e3353/osu.Framework/IO/Stores/ResourceStore.cs#L146-L153), which should make the `textures.Get` lookup in `CreateDefault` work correctly, or at least I would presume?

**Reply by bdach** [reply]
no. it clearly does not work because this discussion exists and there is proof that it doesn't work.

as to why it doesn't work - it doesn't work on the legacy default skin specifically, and is caused by `@2x` suffixes.

https://github.com/ppy/osu/blob/4de66bb1c6e053a948b968d85905c0dda95f8a2b/osu.Game/Skinning/LegacySkin.cs#L483

`LegacySkin` treats component names as extensionless and happily attaches `@2x` to the end of the component name, resulting in lookups like `hitcircle.png@2x`. that one cannot succeed, and neither can the lookup for `hitcircle.png`, because the there is no `hitcircle.png` resource in osu-resources's dll resource store, there's only `hitcircle@2x.png`.

at the point the `CreateDefault()` lookup is taken it's too late to salvage anything. that loads from disk storage (i.e. the `files/` directory).

**Reply by frenzibyte** [reply]
ok, you could make it work by either stripping out the extension in `Storyboard` using `Path.ChangeExtension(path, null)`, or fix the concatenation of `@2x` to handle lookup names with extensions. 

i'm not fussed on either of these given that the first is somewhat of a special case (having an extension in the lookup name), while the second feels more correct...

...but i would say given that `TextureStore`s handle extensions correctly, the second is more likely to be at fault here.

**Comment by Euro20179** [comment]
I give up on trying to upload a video, i have no idea what i'm doing wrong because somehow @Walavouchey was able to post it, i cannot get it under 10mb.

Anyway, imagine @Walavouchey's video, but every time there's a missing approach circle, the circle itself is also missing.
In addition the circle on the first fake slider is missing, and on any slider that is going backwards.
![image](https://user-images.githubusercontent.com/56898436/156903462-3217feb8-faf1-4b7b-afc4-c0af2b3e87c9.png)
![image](https://user-images.githubusercontent.com/56898436/156903468-0a5b0f66-1a4f-4285-b005-febe5c553ef0.png)

Yes i'm playing with background 0%

**Reply by Walavouchey** [reply]
yeah, a lot of skin sprites are missing. i guess i only paid attention to the approach circles since they affected readability the most

**Comment by peppy** [comment, marked as answer]
This should be mostly resolved in the latest release.
